ts/indicators/ema.src.ts

```diff
@@ -0,0 +1,230 @@
+                calEMA: (number|undefined),
+                index: number,
+                SMA: number
+            ): [number, number];
+            public getValues(
+                series: Series,
+                params: EMAIndicatorParamsOptions
+            ): (boolean|IndicatorValuesObject);
+
+        }
+
+        interface EMAIndicatorOptions extends SMAIndicatorOptions {
+            params?: EMAIndicatorParamsOptions;
+        }
+
+        interface EMAIndicatorParamsOptions extends SMAIndicatorParamsOptions {
+            // for inheritance
+        }
+
+        class EMAIndicatorPoint extends SMAIndicatorPoint {
+            public series: EMAIndicator;
+        }
+
+        interface SeriesTypesDictionary {
+            ema: typeof EMAIndicator;
+        }
+
+    }
+}
+
+import U from '../parts/Utilities.js';
+var isArray = U.isArray;
+
+var seriesType = H.seriesType,
+    correctFloat = H.correctFloat;
+
+/**
+ * The EMA series type.
+ *
+ * @private
+ * @class
+ * @name Highcharts.seriesTypes.ema
+ *
+ * @augments Highcharts.Series
+ */
+seriesType<Highcharts.EMAIndicator>(
+    'ema',
+    'sma',
+    /**
+     * Exponential moving average indicator (EMA). This series requires the
+     * `linkedTo` option to be set.
+     *
+     * @sample stock/indicators/ema
+     *         Exponential moving average indicator
+     *
+     * @extends      plotOptions.sma
+     * @since        6.0.0
+     * @product      highstock
+     * @optionparent plotOptions.ema
+     */
+    {
+        params: {
+            /**
+             * The point index which indicator calculations will base. For
+             * example using OHLC data, index=2 means the indicator will be
+             * calculated using Low values.
+             *
+             * By default index value used to be set to 0. Since Highstock 7
+             * by default index is set to 3 which means that the ema
+             * indicator will be calculated using Close values.
+             */
+            index: 3,
+            period: 9 // @merge 14 in v6.2
+        }
+    },
+    /**
+     * @lends Highcharts.Series#
+     */
+    {
+        accumulatePeriodPoints: function (
+            period: number,
+            index: number,
+            yVal: Array<Array<number>>
+        ): number {
+            var sum = 0,
+                i = 0,
+                y: (number|Array<number>) = 0;
+
+            while (i < period) {
+                y = index < 0 ? yVal[i] : yVal[i][index];
+                sum = sum + (y as any);
+                i++;
+            }
+
+            return sum;
+        },
+        calculateEma: function (
+            xVal: (Array<number>|undefined),
+            yVal: Array<Array<number>>,
+            i: number,
+            EMApercent: number,
+            calEMA: (number|undefined),
+            index: number,
+            SMA: number
+        ): [number, number] {
+            var x: number = (xVal as any)[i - 1],
```

Comment:
@bre1470  I'm a bi uncertain if this is correct as xVal could be undefined. But if not it could possibly mean yet another type of `IndicatorValueObject`. Unless there is another way around of course..

Reply:
I think it is actually not possible at the stage of `getValues`, that `xData` of the linkes series is undefined. So this is probably not correct.